### PR TITLE
persist liquidSources table

### PR DESCRIPTION
**Pull request by Heneman**
Fixes https://github.com/DFHack/dfhack/issues/4131

### CODE REVIEW

File: source.lua
Comment (myk002) [inline source.lua:4]: you'll also need to define hooks to load the state when a map is loaded and re-register with repeat-util

Comment (Heneman) [reply source.lua]: Apologies for creating this prematurely, it's been awhile since I've used GitHub and I was just refamiliarizing myself with it! I'll avoid creating `DRAFT/WIP` PRs in the future :sweat_smile: 

I'll be pushing more updates soon

Comment (myk002) [reply source.lua]: no worries! it's fine to create a draft PR. It's a good way to get continuous feedback when trying something new. There's a button for that, though -- you don't have to put DRAFT in the title. I'll click it for you.
image

Comment (Heneman) [reply source.lua]: I appreciate the help!

So as things currently are, I have the Sources persisting between saves as long as the game isn't closed, but closing and relaunching the game causes the Sources to all be removed again.

Do I need to hook into something else? :thinking: 

Also I'm happy to hear any nitpicks and cleanup recommendations for my changes, this is my first endeavor into DFHack and I'm happy to learn how y'all do things.

File: source.lua
Comment (myk002) [inline source.lua:15]: now that I look at it, you might be better off following a model closer to `prioritize`. this tool will be "enabled" if there are sources in the state map.

Comment (Heneman) [reply source.lua]: that looks like it was the golden ticket

**Comment by Heneman** [comment]
persisting but sources will stack instead of overwrite if one is added to a tile that already has one. i imagine we don't want more than once source on a tile so i can add that in, but this _is_ currently functional :tada:

**Comment by Heneman** [comment]
It's working! IT'S WORKING!

File: source.lua
Comment (myk002) [inline source.lua:8]: you can get rid of `sourceId` and use `GLOBAL_KEY` for this as well. no need to have two separate and different IDs

File: source.lua
Comment (myk002) [inline source.lua:13]: unless you're planning on writing unit tests, you don't need this extra level of indirection. you can use `g_sourcess_list` directly.

File: source.lua
Comment (myk002) [inline source.lua:18]: Since I wrote `prioritize` I found a better way to write this. instead of `has_elements(t)` you can write `next(t)` and take advantage of the built-in `next` keyword

File: source.lua
Comment (myk002) [inline source.lua:21]: e.g.
```lua
function isEnabled()
    return next(g_sources_list)
end
```

File: source.lua
Comment (myk002) [inline source.lua:42]: need a `local` here, both for the function and the `new_source` variable

File: source.lua
Comment (myk002) [inline source.lua:33]: should be a `local` function

File: source.lua
Comment (myk002) [inline source.lua:43]: since you already have `formatPos`, this could be
```lua
print(('Adding %d %s to %s'):format(amount, liquid, formatPos(pos)))
```

File: source.lua
Comment (myk002) [inline source.lua:205]: you shouldn't have this problem because your keys are list indices -- prioritize stores a map

File: source.lua
Comment (myk002) [inline source.lua:222]: this is used for unit tests, which you don't have. you can remove this block

File: source.lua
Comment (myk002) [inline source.lua:226]: this if block is redundant with the one below it

File: source.lua
Comment (myk002) [inline source.lua:232]: this should be inside the if block with `main` since its purpose is to persist any state changes that `main` made

File: source.lua
Comment (myk002) [inline source.lua:109]: since you're using `table.insert` to populate the table, it's a list, not a map. using `ipairs` instead of `pairs` here (and elsewhere in this file) would be semantically clearer.

**Review by myk002 — COMMENTED** [review]
looks like all the major pieces are in place! great job! just needs some cleanup. Thank you for getting this done! I can do another review pass once this is cleaned up.

**Comment by Heneman** [comment]
pre-commit.ci autofix

**Comment by Heneman** [comment]
o.O 

I thought I'd be triggering a cleanup of the trailing whitespace that the precommit check was showing failed, I didn't intend for a build to be triggered. Apologies for any mess that caused :grimacing:

**Comment by myk002** [comment]
> I didn't intend for a build to be triggered. Apologies for any mess that caused

A build is *supposed* to be triggered after every change -- that's how we detect potential failures before we merge them. Not a mess. Working as intended.

File: source.lua
Comment (myk002) [inline source.lua:96]: neither of these lines should be necessary. you'll be persisting state later, and you don't need to register the repeat when deleting a source from the list

File: source.lua
Comment (myk002) [inline source.lua:118]:
```suggestion
    for _,v in ipairs(g_sources_list) do
```

File: source.lua
Comment (myk002) [inline source.lua:189]: you should also clear your state when a world is unloaded

**Comment by Heneman** [comment]
Should be ready for :eyes: again! The rearrangement of functions and some cleanup means some of the comments aren't resolving, but I _think_ I got everything that was active

File: source.lua
Comment (myk002) [inline source.lua:70]:
```suggestion
```

Comment (myk002) [reply source.lua:70]: no need to register with repeat-util when removing a source

File: source.lua
Comment (myk002) [inline source.lua:181]: this needs to go above the other if statement, otherwise it will never be run

**Review by myk002 — COMMENTED** [review]
looks good! remember to add a line to the Misc Improvements section (under the "Future" version header) of [changelog.txt](https://github.com/DFHack/scripts/blob/master/changelog.txt#L42C2-L42C2). Something like:

```
- `source`: now saves its state with the fort
```

**Review by myk002 — APPROVED** [review]

**Comment by myk002** [comment]
Great job with this! Thank you for working with me through all the comments!

**Comment by Heneman** [comment]
Happy to help! 

Headed back to my failed megaproject with my newfound magic powers:

!(so long)[https://media4.giphy.com/media/gIN6MT17NtapJqFr25/giphy.gif]

**Pull request merged**
